Sources/Workspace/WorkspaceConfiguration.swift

```diff
@@ -350,30 +350,13 @@ extension Workspace.Configuration {
         }
 
         private static func load(_ path: AbsolutePath, fileSystem: FileSystem) throws -> AuthorizationProvider {
```

Comment:
Workspace.Configuration.Netrc is designed to only concern itself with `NetrcAuthorizationProvider`, so most of the code that was removed here should be put back, and then in `SwiftTool::getAuthorizationProvider ` we would create the `CompositeAuthorizationProvider` from the one returned from `self.getNetrcConfig` + the keychain one, i.e.

```
func getAuthorizationProvider() throws -> AuthorizationProvider? {
        // currently only single provider: netrc
        return try self.getNetrcConfig()?.get()
    }
```

Reply:
I undid the change in `load`, but `NetrcAuthorizationProvider` can be generic?

Quoting [SE-0292](https://github.com/apple/swift-evolution/blob/main/proposals/0292-package-registry-service.md#specifying-credentials-for-a-custom-registry):

> If the user passes the --login and --password options to the set subcommand along with the --global option, the user-level .netrc file is updated instead. When Swift Package Manager connects to a custom registry, it first consults the project's .netrc file, if one exists. If no entry is found for the custom registry, Swift Package Manager then consults the user-level .netrc file, if one exists.

So potentially in `SwiftTool::getAuthorizationProvider` we'd need to add another `NetrcAuthorizationProvider` that points to the user-level netrc.

Reply:
Amended https://github.com/apple/swift-package-manager/pull/3768/commits/6cc7ece871f211d1fde25e065372cf111739bd51v

Reply:
1. making NetrcAuthorizationProvider make sense,

2. I _think_ the local vs. user netrc file logic could go into

```
func getNetrcConfig() -> Workspace.Configuration.Netrc? {
        guard options.netrc else { return nil }

        if let configuredPath = options.netrcFilePath {
            guard localFileSystem.exists(configuredPath) else {
                self.observabilityScope.emit(error: "Did not find .netrc file at \(configuredPath).")
                return nil
            }

            return .init(path: configuredPath, fileSystem: localFileSystem)
        } else {
            let defaultPath = localFileSystem.homeDirectory.appending(component: ".netrc")
            guard localFileSystem.exists(defaultPath) else { return nil }

            return .init(path: defaultPath, fileSystem: localFileSystem)
        }
    }
```

which now loads the netrc by either a flag passed to the CLI or the user level, but this can be extended to also include the local case ?

Reply:
I actually think that with this change Workspace.Configuration.Netrc becomes redundant? we can add a throwing initializer to `NetrcAuthorizationProvider` to load the file and remove it?

Reply:
> any reason why not simply remove it? 

My original intent with this PR is to add keychain auth capability. I didn't plan to change existing behavior or cover new ones.

Reply:
> I actually think that with this change Workspace.Configuration.Netrc becomes redundant?

`Workspace.Configuration.Netrc` might be used elsewhere through libSwiftPM? It doesn't feel right to me that only `SwiftTool` needs this logic.

> I think the local vs. user netrc file logic could go into

That code is for `SwiftTool` only and I assume `SwiftPackageRegistryTool` will have to create its own `AuthorizationProvider` in a similar way. I don't know if the logic for both tools is supposed to be the same, but my current understanding is that they are not. 

AFAICT `options.netrcFilePath` isn't necessarily a path within the workspace, so `SwiftTool` uses either a custom netrc file location, which can be anywhere, or tries to find one in user's home directory. Regardless, it uses only one netrc file. 

`SwiftPackageRegistryTool` IIUC, looks for netrc file in both the workspace (same as project?) and user's home directory. i.e., it might use two netrc files.

If the tools were to share the same logic, then the code should be moved outside of `SwiftTool` (to where?).

Reply:
SwiftTool is a shared impl for all the CLI tools, so code there would be shared across them, but you are right that this is not needed to be dealt with in this PR - we can follow up with consolidating the logic as we make more progress